### PR TITLE
Add HTTP URLs verification by default

### LINKED ISSUE #11: Only allow HTTP URLs by default
```js
isAbsoluteUrl('javascript:throw%20document.cookie') // true
```

Although this might be technically correct (?), it seems like it's not what most people would think of as an absolute URL, and it could lead to a false sense of security that the input has been validated as a 'real' URL.

Possible solutions:

- Release a new major version that ensures the protocol is `http` or `https`
  - Maybe with an option like `{ allowNonHttp: true }` for people who want the old behaviour.
- Or just add a warning in the readme that this will return true for any protocol, even `javascript:`.

### PR DESCRIPTION
**Pull request by JoaoOliveiraRocha**
Attempt to cover #11. Possible breaking changes.

It now verifies if it's an HTTP URL using a RegEx based on [this](https://stackoverflow.com/a/3809435)

You can now use `isAbsolutePath('http://sindresorhus.com')`. It will behave as `isAbsolutePath('http://sindresorhus.com', true)`.

There is possible to `isAbsolutePath('http://sindresorhus.com', false)` and get the same results as before this PR (it even uses the original RegEx to evaluate the string).

I've decided to K.I.S.S. and keep the PR as you mentioned here, in #11 and #12. Since there are no requests for extra features there's no need to "prepare" for expansion right now.

**Note:** this package returns `true` on situations like `http://sindr esorhus.com` (white-space in between). I kept its original behavior. If you think it would be helpful for this package to check for whitespaces (and warn/return false) please inform me!

~~This last object, called Options, can have some parameters (so it's easier to add extra features later). Right now I've only done 'httpOnly'.~~

~~It has 3 levels of 'strictness': `'strictest'`, `'strict'` and `'loose'` / `'true`. Basically `strictest` is the rigid `http://www.lorem.ipsum.com/path`, the `strict` is something your browser could understand, and `loose` only checks for the HTTP protocol header.~~

~~I have wrote some examples on the /readme.md file.~~

~~I've also copied the current array of tests to test against the 'levels of strictness'. If you have any doubt about each link I suggest open the test files and see what is the result of each "URL" against each parameter.~~

---

Fixes #11

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -1,19 +1,31 @@
...
+  
+
 ```
+
 $ npm install is-absolute-url
+
 ```
 
+  
+
```
Comment (sindresorhus) [inline readme.md:21]: Don't do unrelated changes. Applies everywhere.

File: index.d.ts [hunk 2]
```diff
@@ -17,6 +17,6 @@ isAbsoluteUrl('foo/bar');
 //=> false
 ```
 */
-declare function isAbsoluteUrl(url: string): boolean;
+declare function isAbsoluteUrl(url: string, options: object): boolean;
```
Comment (sindresorhus) [inline index.d.ts:20]: This should be strongly-typed. See: https://github.com/sindresorhus/typescript-definition-style-guide/tree/57a925bb5b3eb46a4666eab31e1c883a26b7bfb5#documentation Should also have docs.

**Comment by sindresorhus** [comment]
There are a lot of typos and formatting problems. Try to look over the diff and improve it: https://github.com/sindresorhus/is-absolute-url/pull/14/files

**Comment by sindresorhus** [comment]
Why 3 states instead of just a `httpOnly` option as mentioned in https://github.com/sindresorhus/is-absolute-url/issues/11?

**Comment by JoaoOliveiraRocha** [comment]
> Why 3 states instead of just a `httpOnly` option as mentioned in #11?

It's possible to simply set `httpOnly` as `true` and it will behave the way #11 intended. 

I've added the 3 states because 1) according to the RFC: URLs should be case-sensitive. And 2) it could be expanded in the future to check other conditions p.e. if is an absolute ftp address, contains the port or to check for secure `HTTPS` only ( p.e. pass `{ httpOnly: true, portOnly: true }` for validating a web link with a defined port ).

Or do you think it's better to keep it simple and go in a way #12 tried to archive?

File: index.js [hunk 3]
```diff
@@ -1,16 +1,35 @@
 'use strict';
 
-module.exports = url => {
+// Options - httpOnly
```
Comment (satazor) [inline index.js:3]:
```suggestion
// Options
// - httpOnly
```

File: readme.md [hunk 4]
```diff
@@ -26,19 +38,90 @@ isAbsoluteUrl('foo/bar');
 ```
 
 
-## Related
+## Options Parameters
```
Comment (satazor) [inline readme.md:41]:
```suggestion
## Options
```

File: readme.md [hunk 5]
```diff
@@ -26,19 +38,90 @@ isAbsoluteUrl('foo/bar');
...
+`isAbsoluteUrl('httpS://www.sindresorhus', {httpOnly: 'strict')` => `false`  
+`isAbsoluteUrl('httpS://www.sindresorhus', {httpOnly: 'loose')` => `true`
+
+This way it's possible to quickly differentiate a string that starts with the `http` protocol vs a "proper" absolut Web Url.
+
+#### > 'strictest'
+Checks for the `https://` (or non-secure `http`), `www` and the TLDomain (p.e. `.com`)
+
+```js
+isAbsoluteUrl('https://www.sindresorhus.com/path/to/file', {httpOnly: 'strictest');
```
Comment (satazor) [inline readme.md:60]:
```suggestion
isAbsoluteUrl('https://www.sindresorhus.com/path/to/file', {httpOnly: 'strictest'});
```

Comment (satazor) [reply readme.md:60]: You are missing a `}` in all examples below as well.

File: readme.md [hunk 6]
```diff
@@ -26,19 +38,90 @@ isAbsoluteUrl('foo/bar');
...
+  
+  
+
+## Related
+
+  
+
+See [is-relative-url](https://github.com/sindresorhus/is-relative-url) for the inverse.
 
-<div align="center">
```
Comment (satazor) [inline readme.md:36]: Why was this removed?

**Review by satazor — CHANGES_REQUESTED** [review]
There are a lot of unrelated changes in the docs.

File: readme.md [hunk 7]
```diff
@@ -26,19 +38,90 @@ isAbsoluteUrl('foo/bar');
 ```
 
 
-## Related
+## Options Parameters  
 
-See [is-relative-url](https://github.com/sindresorhus/is-relative-url) for the inverse.
+### httpOnly  
+
```
Comment (satazor) [inline readme.md:44]: I would keep the docs simple like how it's done in the Usage section. A simple js block with lines and their result as comments would be better for readability.

File: readme.md (hunk 6 above)
Comment (JoaoOliveiraRocha) [reply readme.md]: crtl-c + crtl-v mistakes

File: readme.md (hunk 7 above)
Comment (JoaoOliveiraRocha) [reply readme.md]: Understood. Can you provide feedback now please?

File: readme.md (hunk 5 above)
Comment (JoaoOliveiraRocha) [reply readme.md]: Ups... fixed it

File: index.d.ts [hunk 8]
```diff
@@ -17,6 +17,10 @@ isAbsoluteUrl('foo/bar');
 //=> false
 ```
 */
-declare function isAbsoluteUrl(url: string): boolean;
+interface Options {
```
Comment (sindresorhus) [inline index.d.ts:20]: This needs to be inside a namespace: https://github.com/sindresorhus/typescript-definition-style-guide

File: index.d.ts [hunk 9]
```diff
@@ -17,6 +17,10 @@ isAbsoluteUrl('foo/bar');
...
+    httpOnly?: boolean | string;
```
Comment (sindresorhus) [inline index.d.ts:21]: Tab-indentation

File: index.d.ts [hunk 10]
```diff
@@ -1,7 +1,7 @@
 /**
 Check if a URL is absolute.
-
 @param url - The URL to check.
+@param options - Object with different parameters to check.
```
Comment (sindresorhus) [inline index.d.ts:4]: This doesn't add much. I would remove it. People know what "options" means.

File: index.d.ts [hunk 11]
```diff
@@ -1,7 +1,7 @@
 /**
 Check if a URL is absolute.
-
```
Comment (sindresorhus) [inline index.d.ts:3]: Don't do unrelated changes.

File: readme.md [hunk 12]
```diff
@@ -41,4 +56,4 @@ See [is-relative-url](https://github.com/sindresorhus/is-relative-url) for the i
 	<sub>
 		Tidelift helps make open source sustainable for maintainers while giving companies<br>assurances about security, maintenance, and licensing for their dependencies.
 	</sub>
-</div>
+</div>
```
Comment (sindresorhus) [inline readme.md:59]: Don't do unrelated changes.

**Comment by sindresorhus** [comment]
I think it should be a single `httpOnly` boolean option.

**Comment by sindresorhus** [comment]
>  - strictest	> Everything accordingly to RFC 1738 -  Case sensititive

This is incorrect. The URL scheme is case-insensitive.

File: index.test-d.ts [hunk 13]
```diff
@@ -1,4 +1,4 @@
 import {expectType} from 'tsd';
 import isAbsoluteUrl = require('.');
 
-expectType<boolean>(isAbsoluteUrl('https://sindresorhus.com/foo/bar'));
+expectType<boolean>(isAbsoluteUrl('https://sindresorhus.com/foo/bar', { httpOnly: true }));
```
Comment (sindresorhus) [inline index.test-d.ts:4]:
```suggestion
expectType<boolean>(isAbsoluteUrl('https://sindresorhus.com/foo/bar', {httpOnly: true}));
```

**Comment by sindresorhus** [comment]
You missed this from the original issue:

> How about an option called httpOnly **which defaults to true**

File: readme.md [hunk 14]
```diff
@@ -25,6 +25,21 @@ isAbsoluteUrl('foo/bar');
...
+
+### httpOnly
+
+You can pass an object as a second parameter.
+
+```js
+isAbsoluteUrl('https://si', { httpOnly: true });
+//=> true
+
+is absoluteUrl('mailto:someone@example.com', { httpOnly: true });
```
Comment (sindresorhus) [inline readme.md:38]: Typo

File: readme.md [hunk 15]
```diff
@@ -25,6 +25,21 @@ isAbsoluteUrl('foo/bar');
 //=> false
 ```
 
+## Options
```
Comment (sindresorhus) [inline readme.md:28]: Use the same format for readme docs as used here: https://github.com/sindresorhus/boxen#api

File: test.js [hunk 16]
```diff
@@ -17,3 +18,138 @@ test('main', t => {
...
+	t.false(isAbsoluteUrl('//sindresorhus.com', {}));
+	t.false(isAbsoluteUrl('/foo/bar', {}));
+	t.false(isAbsoluteUrl('foo/bar', {}));
+	t.false(isAbsoluteUrl('foo', {}));
+	t.false(isAbsoluteUrl('c:\\', {}));
+	t.false(isAbsoluteUrl('c:\\Dev\\test-broken', {}));
+	t.false(isAbsoluteUrl('C:\\Dev\\test-broken', {}));
+	t.false(isAbsoluteUrl('ht,tp://sindresorhus.com', {}));
+});
+
```
Comment (sindresorhus) [inline test.js:155]:
```suggestion
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
The pull request needs a proper descriptive title.

**Comment by JoaoOliveiraRocha** [comment]
> You missed this from the original issue:
> 
> > How about an option called httpOnly **which defaults to true**

Done. I've also changed the title and description of this PR.

File: test.js [hunk 17]
```diff
@@ -17,3 +17,61 @@ test('main', t => {
 	t.false(isAbsoluteUrl('C:\\Dev\\test-broken'));
 	t.false(isAbsoluteUrl('ht,tp://sindresorhus.com'));
 });
+
+test('HttpOnly set as true (same as default behavior). Assert TRUE tests', t => {
```
Comment (sindresorhus) [inline test.js:21]: Too verbose test title. Formatting needs work too.

File: test.js [hunk 18]
```diff
@@ -17,3 +17,61 @@ test('main', t => {
...
+	t.true(isAbsoluteUrl('file://sindresorhus.com', false));
+	t.true(isAbsoluteUrl('mailto:someone@example.com', false));
+	t.true(isAbsoluteUrl('data:text/plain;base64,SGVsbG8sIFdvcmxkIQ%3D%3D', false));
+	t.true(isAbsoluteUrl('https://www.sindresorhus.com', false));
+	t.true(isAbsoluteUrl('https://www.sindres.orhus.com/path', false));
+	t.true(isAbsoluteUrl('httpS://www.sindresorhus.com', false));
+	// eslint-disable-line no-script-url
+	t.true(isAbsoluteUrl('javascript:throw%20document.cookie', false));
+
+	t.true(isAbsoluteUrl('https://sindresorhus', false));	// Should this behavior exist? missing TLD?...
```
Comment (sindresorhus) [inline test.js:63]: Instead of asking questions, try to research and figure it out.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: test.js (hunk 18 above)
Comment (JoaoOliveiraRocha) [reply test.js]: I felt it should at least check for the TLD, but it would return false to URLs as `http://localhost:7700` and they are absolute. So it will not check for it.

File: index.js [hunk 19]
```diff
@@ -1,10 +1,14 @@
 'use strict';
 
-module.exports = url => {
+module.exports = (url, options = {httpOnly: true}) => {
```
Comment (sindresorhus) [inline index.js:3]: This is not the correct way to do default values for options objects. If another option is added, it will silently not do the right thing as both options will be overwritten if one is set.

```suggestion
module.exports = (url, {httpOnly = true} = {}) => {
```

File: index.js [hunk 20]
```diff
@@ -1,10 +1,14 @@
...
 	if (typeof url !== 'string') {
 		throw new TypeError(`Expected a \`string\`, got \`${typeof url}\``);
 	}
 
+	if (options.httpOnly) {
+		return /^((http[s]?)?:\/\/([-a-z0-9()@:%_.~#?&//=]*))/gi.test(url);
```
Comment (sindresorhus) [inline index.js:9]: The regex could still be simplified.

File: readme.md [hunk 21]
```diff
@@ -25,6 +25,26 @@ isAbsoluteUrl('foo/bar');
...
 ```
 
+### Options
+
+#### httpOnly
+
+Type: `boolean`\
+Default: `true`
+
+Check for the HTTP protocol in a URL.
```
Comment (sindresorhus) [inline readme.md:35]: Still not clear enough.
```suggestion
Only accept absolute HTTP URLs.
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]
